src/module_4/push_exploration_non_linear.md

```diff
@@ -0,0 +1,2005 @@
+):
+    """
+    Generate ROC and Precision-Recall curves for multiple binary classification models
+    and display them in a single figure.
+    """
+
+    # Create a timestamp for unique filenames
+    timestamp = datetime.now().strftime("%Y%m%d%H%M%S")
+
+    plt.figure(figsize=(10, 5))
+
+    # Plot for each model
+    for model_name, y_pred in zip(model_names, y_preds):
+
+        # ROC Curve
+        fpr, tpr, _ = roc_curve(y_test, y_pred)
+        roc_auc = auc(fpr, tpr)
+
+        plt.subplot(1, 2, 1)
+        plt.plot(fpr, tpr, label=f"ROC (AUC = {roc_auc:.2f}) - {model_name}")
+        
+        # Precision-Recall Curve
+        precision, recall, _ = precision_recall_curve(y_test, y_pred)
+        pr_auc = auc(recall, precision)
+
+        plt.subplot(1, 2, 2)
+        plt.plot(recall, precision, label=f"PR (AUC = {pr_auc:.2f}) - {model_name}")
+
+    # ROC Curve settings
+    plt.subplot(1, 2, 1)
+    plt.plot([0, 1], [0, 1], color="navy", linestyle="--")
+    plt.xlim([0.0, 1.0])
+    plt.ylim([0.0, 1.05])
+    plt.xlabel("False Positive Rate")
+    plt.ylabel("True Positive Rate")
+    plt.title("ROC")
+    plt.legend(loc="lower right", prop={'size': 6})
+    
+    # Precision-Recall Curve settings
+    plt.subplot(1, 2, 2)
+    plt.xlim([0.0, 1.0])
+    plt.ylim([0.0, 1.05])
+    plt.xlabel("Recall")
+    plt.ylabel("Precision")
+    plt.title("Precision-Recall Curve")
+    plt.legend(loc="upper right", prop={'size':6})
+
+    plt.tight_layout()
+
+    if save_curves_path:
+        # Define the filename with a timestamp
+        figure_filename = f"Evaluation_Curves_Multiple_Models_{timestamp}.png"
+        figure_path = os.path.join(save_curves_path, figure_filename)
+
+        plt.savefig(figure_path)
+
+    plt.show()
+```
+
+
+```python
+# Train regression tree
+tree = DecisionTreeClassifier()
+tree.fit(X_train, y_train)
+y_pred_tree = tree.predict(X_val)
+
+# Train logistic regression
+lr = LogisticRegression(penalty="l2", C=1e-5, solver="liblinear")
+lr.fit(X_train, y_train)
+y_pred_lr = lr.predict_proba(X_val)[:, 1]
+
+generate_evaluation_curves_multi(
+    model_names=["ClassifierTree", "LogisticRegression"],
+    y_preds=[y_pred_tree, y_pred_lr],
+    y_test= y_val
+)
+
+```
+
+
+    
+![png](push_exploration_non_linear_files/push_exploration_non_linear_14_0.png)
+    
+
+
+Obviously, a single decision tree will hardly beat it. So we will try using random forests with different params of n_trees. As as curiosity, if we take a look at the plot for LR with same parameters as trained before, the AUC for PR is 0.16 vs 0.17 before (C, solver and  regularisation remains equal). I think this might happen as model regularisation can still have some randomness in model weights, hence getting different results. For addressing that we could have made use of random_seed.
+
+## Random Forests
+
+
+```python
+# Train Logistic Regression
+lr = LogisticRegression(penalty="l2", C=1e-5, solver="liblinear")
```

Comment:
Important homework: you are using l2 penalty (ridge) to overcome overfitting, but you are never checking if that's the case: for that compare performance in train and test. Play with the C parameter to see how overfitting is reduced and bias is increased